### DISCUSSION TITLE
orderBy _relevance appears to fall back to a default sort parameter instead

### DISCUSSION THREAD

**Discussion by AlanHarrisonPeak** [prisma/prisma#14744] [Q&A]
Hi, 
 Prisma v3.15.0, Postgres v13.5, node v14.19, 

Im attempting to order filtered results by relevance to search term based on the "name" column in my db. There is, for reference, also "createdAt" and "tenant_name", "featured" and "id" columns.

Im having the issue where these following 2 records are not returning in expected order of relevance to the search term "the-r" -  I would expect the order to be opposite to that displayed below. Not only that, but if i manually edit the createdAt and invert them, the order reverses, indicating _relevance is not working. I have also run the raw SQL using DBeaver and the same issue occurs, leading me to wonder if either im not fully understanding the syntax, or the syntax doesnt work: 

query params are `
{ 
   featured: false,
   name: 'the-r',
   tenant: 'mock-tenant'
}
`

the result is : 

``` 
{
  "pageCount": 1,
  "pageNumber": "1",
  "pageSize": "25",
  "specCount": 2,
  "specs": [
    {
      "createdAt": "2022-07-21T10:05:03.172Z",
      "featured": false,
      "id": "a0ff1c71-f39f-4125-ac6e-c84aa6aebaaf",
      "name": "the-fellowship-of-the-ring",
      "tenant": "mock-tenant"
    },
    {
      "createdAt": "2022-07-20T10:07:32.498Z",
      "featured": true,
      "id": "46ba562d-231d-4baa-b8b9-386a89a36f1c",
      "name": "the-return-of-the-king",
      "tenant": "mock-tenant"
    }
  ]
} 
```

Im using `previewFeatures = ["referentialIntegrity", "filterJson", "fullTextSearch"]`

Im using this query : 

``` 
db.specs.findMany({
          orderBy: {
            ...(name
              ? {
                  _relevance: {
                    fields: ['name'],
                    search: name,
                    sort: 'desc',
                  },
                }
              : { created_at: 'desc' }),
          },
          skip: queryOffset,
          take: queryLimit,
          where: {
            featured: featured ? true : {},
            name: name
              ? {
                  contains: name,
                  mode: 'insensitive',
                }
              : {},
            tenant_name: tenant,
          },
        }), 
```

which outputs this SQL :

``` 
SELECT * FROM "public"."specs" WHERE ("public"."specs"."name" ILIKE $1 AND "public"."specs"."tenant_name" = $2) ORDER BY ts_rank(to_tsvector(concat_ws(' ', "public"."specs"."name")), to_tsquery($3)) DESC LIMIT $4 OFFSET $5
 ``` 

any help would be greatly appreciated!

**Comment by do4gr** [comment]
Hey, thanks for opening this discussion. 
I'm trying to understand whether there is a problem. Why do you expect a specific order between these records? Both contain the substring `the-r` and logging the rank that ts_rank creates shows them to have the same rank.

```sql
SELECT *, rank 
FROM "public"."specs",  
ts_rank(to_tsvector(concat_ws(' ', "public"."specs"."name")), to_tsquery($1)) as rank
WHERE ("public"."specs"."name" ILIKE $2 AND "public"."specs"."tenant_name" = $3) 
ORDER BY rank DESC 
LIMIT $4 OFFSET $5
```

The query we generate ranks by number of occurences of the search term and does not use any form of normalization so I think it's expected that both rows rank the same. (See 12.3.3 [here](https://www.postgresql.org/docs/current/textsearch-controls.html)).
Would you expect some form of normalization (like the optional modes available on ts_rank that you can find in the linked PG docs)?

**Reply by AlanHarrisonPeak** [reply]
I am also trying to understand 😄  I expected ts_rank to give more weight to the string that BEGINS with the search term over those that include, as per [that same doc](https://www.postgresql.org/docs/current/textsearch-controls.html) 12.3.3 (see reference to "neutrinos") - the first table is un-normalised and still returns the record that BEGINS WITH having a higher rank. The normalised table below it has them ranked closer, oddly. Maybe im misunderstanding the docs though, so I will read through your optionals and see if there is something to suit. thanks!

**Reply by AlanHarrisonPeak** [reply]
Ah, i see the issue. The table im referring to is ts_rank_cd, and tbh that is what i would expect to be the default behavior. To achieve this in prisma, will I have to use a $rawQuery to modify the behavior? if so, could be worth starting to add these modifiers to your syntax.

**Reply by do4gr** [reply]
Yes, currently we default to ts_rank and don't offer to configure that in the API. So $raw is the only way to achieve that currently. We're thinking already about enabling the choice of query and ranking modes. We'll see whether we can come up with a nice API for this.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
